File: go-algorithm/leetcode/problem75.go
package leetcode

// 75. Sort Colors
// https://leetcode.com/problems/sort-colors/

func sortColors(nums []int) {
	i, l, r := 0, 0, len(nums)-1
	for i <= r {
		if nums[i] == 0 {
			nums[i], nums[l] = nums[l], nums[i]
			l++
			i++
		} else if nums[i] == 2 {
			nums[i], nums[r] = nums[r], nums[i]
			r--
		} else {
			i++
		}
	}
}

File: java-algo/src/main/java/com/leetcode/Problem75.java
package com.leetcode;

/**
 * 75. Sort Colors
 * https://leetcode.com/problems/sort-colors/
 */
public class Problem75 {
    public void sortColors(int[] nums) {
        int[] cnt = new int[3];
        for (int num : nums) {
            cnt[num]++;
        }
        int i = 0, j = 0;
        while (i < nums.length) {
            while (j < 3 && cnt[j] == 0) {
                j++;
            }
            if (j == 3) {
                break;
            }
            nums[i++] = j;
            cnt[j]--;
        }
    }
}
